tests/common/fixtures/elasticsearch.py

```diff
@@ -22,6 +22,10 @@ def init_elasticsearch(request):
     """Initialize the test Elasticsearch index once per test session."""
     client = _es_client()
 
+    from h.search.persistence import Annotation
+    Annotation._doc_type.index = "hypothesis-test"
+    Annotation.init()
```

Comment:
Here I'm hacking `Annotation._doc_type.index` to get the tests to use a different index than dev.

Also creating and initializing the index.

Reply:
Eew ... this is a bit ugly. It means that someone reading the code for `h/search/persistence.py` is not going to know straight away that a different index is used in the test, since the index value is not injected.

I usually prefer to inject dependencies in an explicit manner to avoid this confusion. For that reason I'm not super enthusiastic about the connection singleton either.

I'm wondering whether we could at least determine the index name in `h.search.persistence` rather than monkey-patching it elsewhere. In tox we could use environment variables for this presumably?

Reply:
http://elasticsearch-dsl.readthedocs.io/en/latest/persistence.html#index. I think the more official way to do this is to create an index and then register a doctype with it so:
```
hypothesis_test = Index('hypothesis-test')
hypothesis_test.doc_type(Annotation)
hypothesis_test.create()
```

Reply:
It certainly looks simpler to call `Annotation.init()` and that's what the docs suggest: http://elasticsearch-dsl.readthedocs.io/en/latest/persistence.html#document-life-cycle The docs about the more complex `Index`-based approach suggest that approach is useful when creating multiple mappings at once or when doing migrations.

Reply:
To Rob's eew ... What I think we should do here is have the top-level search indexing and querying callables (`index()`, `delete()`) take an optional `index` argument with a hard-coded default `index="hypothesis"`. So the code won't actually need to pass `index` since it'll use the hardcoded default. The _tests_ will need pass `index="hypothesis-test"` on every call but the tests can use an `index()` wrapper fixture to make this invisible.

So this will mean that the `Annotation` class would _not_ have the hardcoded `Annotation.Meta.index = "hypothesis"` anymore (the `class Meta` inside `Annotation` can be deleted) and then whenever the `h.search` code creates an annotation it has to do:

```python
def index(..., index="hypothesis"):
    annotation = Annotation(meta={"index": index})
    ...
```

Reply:
Hi, folks! I've spent a bit of time trying different approaches to injecting `index`. I also find the `Meta` index hardcoding-then-overriding on the "ewww" side. Buuuutttt....

I want to caution that there will be a not-inconsiderable amount of index-name-juggling in the immediate future—_especially in tests—, at least until we get more wrappers around the exposed API of search. I'm a little wary that the cure is worse than the disease in this particular case. And it may force us to wrap things that really shouldn't need to be wrapped (especially per test needs).

One needs to have access to the correct index name to use anytime one:

* Initializes a (Annotation) doctype (via the class method `Annotation::init`)
* Creates a new annotation (via the class method `Annotation::create`)
* Retrieves an annotation (via class method `Annotation::get`)
* Instantiates an Annotation (alluded to in previous comment)
* etc. etc.

Yes, we can wrap a number of these interactions, but...again...this is a bunch of work. I'm already looking at a rather ungainly diff in the little bits of this I've implemented so far. TBH, I'm surprised at how bleh it feels as I was in spiritual agreement with @seanh 's suggested direction.

I want to get a reading on how passionately folks feel about this `Meta` magic—given that I think we're looking at a not-inconsiderable amount of contortion to get around it. I may be approaching this wrongly; at present it is feeling quite fidgety.

Reply:
I share @lyzadanger 's concern about having to pass in index everywhere. I think just like the default connection it would be useful to have a default index and I think that's kinda the idea behind the meta index?? I don't feel super strong one way or the other as I think they will both work but if we're going for simple from a calling es perspective and since the index as @seanh pointed out is the same throughout our code, it seems like you would want to take advantage of the ability to default/set the index once and not worry about it.

Reply:
Here's kinda a related github issue on the subject: https://github.com/elastic/elasticsearch-dsl-py/issues/746.

Reply:
Hmm. Tests doing `Annotation._doc_type.index = "hypothesis-test"` does seem hacky (I don't know if elasticsearch_dsl provides a less hacky-seeming way to do it?) but it's one line, and only runs in the tests doesn't affect production code at all. Do the benefits outweigh the costs, if it avoids an index argument everywhere?

We may run into this again when implementing re-indexing

Reply:
> (I don't know if elasticsearch_dsl provides a less hacky-seeming way to do it?)

According to the author no (see the github issue I linked to), you can either use the _doc_type or not set a default and pass it in every time. It sounds like testing on a different index was not considered. In regards to re-indexing I think you would use the method for index migration in the docs. (Similar to the code I posted earlier.)

Reply:
> Do the benefits outweigh the costs, if it avoids an index argument everywhere?

I would say yes in this case they do but that's just me. :)

Reply:
I'm happy with whatever @lyzadanger decides on this in the end, though